packages/core/src/anchor/processing-loop.ts

```diff
@@ -100,28 +110,41 @@ export class ProcessingLoop<T> {
       this.#abortController.signal.addEventListener('abort', done)
     })
     const processing = async (): Promise<void> => {
-      try {
-        let isDone = false
-        do {
+      let taskId = 0
+      let isDone = false
+      do {
+        try {
           this.#logger.verbose(`ProcessingLoop: Fetching next event from source`)
           const next = await Promise.race([this.source.next(), rejectOnAbortSignal])
           isDone = next.done
           if (isDone) break
           const value = next.value
-          if (!value) {
+          if (value === undefined) {
             this.#logger.verbose(`No value received in ProcessingLoop, skipping this iteration`)
             continue
           }
-          await Promise.race([this.handleValue(value), rejectOnAbortSignal])
-        } while (!isDone)
-        this.#whenComplete.resolve()
-        this.#logger.debug(`ProcessingLoop complete`)
-      } catch (e) {
-        this.#logger.err(`Error in ProcessingLoop: ${e}`)
-        this.#whenComplete.reject(e)
-      }
+          const curTaskId = taskId
+          const task = this.#semaphore
+            .use(async () => {
+              await Promise.race([this.handleValue(value), rejectOnAbortSignal])
+            })
+            .catch((e) => {
+              this.#logger.err(`Error in ProcessingLoop: ${e}`)
+              this.#abortController.abort('ERROR')
+            })
+            .finally(() => {
+              this.#runningTasks.delete(curTaskId)
+            })
+          this.#runningTasks.set(taskId++, task)
```

Comment:
This feels incorrect. Beyond just number overrun, this functionality should be run in one of the available 'slots' with the task pulled into a slot as others complete.

Reply:
Yeah, unfortunately no other approach I could think of allowed a task to easily remove itself from the set on completion.  For now I just modded the taskId by Number.MAX_SAFE_INTEGER to handle wrap-around